nft_ingester/src/program_transformers/candy_machine/candy_machine.rs

```diff
@@ -0,0 +1,220 @@
+
+    let last_minted = if let Some(candy_machine_model) = candy_machine_model {
+        if candy_machine_model.items_redeemed < candy_machine.items_redeemed as i64 {
+            Some(Utc::now())
+        } else {
+            candy_machine_model.last_minted
+        }
+    } else {
+        None
+    };
+
+    let candy_machine_state = candy_machine::ActiveModel {
+        id: Set(id.0.to_vec()),
+        authority: Set(candy_machine.authority.to_bytes().to_vec()),
+        wallet: Set(Some(candy_machine.wallet.to_bytes().to_vec())),
+        token_mint: Set(token_mint),
+        items_redeemed: Set(candy_machine.items_redeemed as i64),
+        version: Set(2),
+        created_at: Set(Some(Utc::now())),
+        last_minted: Set(last_minted),
+        ..Default::default()
+    };
+
+    let query = candy_machine::Entity::insert(candy_machine_state)
+        .on_conflict(
+            OnConflict::columns([candy_machine::Column::Id])
+                .update_columns([
+                    candy_machine::Column::Authority,
+                    candy_machine::Column::Wallet,
+                    candy_machine::Column::TokenMint,
+                    candy_machine::Column::ItemsRedeemed,
+                    candy_machine::Column::LastMinted,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    txn.execute(query)
+        .await
+        .map(|_| ())
+        .map_err(|e: DbErr| IngesterError::DatabaseError(e.to_string()))?;
+
+    let (mode, presale, whitelist_mint, discount_price) =
+        if let Some(whitelist) = data.whitelist_mint_settings {
+            let mode = match whitelist.mode {
+                blockbuster::programs::candy_machine::state::WhitelistMintMode::BurnEveryTime => {
+                    WhitelistMintMode::BurnEveryTime
+                }
+                blockbuster::programs::candy_machine::state::WhitelistMintMode::NeverBurn => {
+                    WhitelistMintMode::NeverBurn
+                }
+            };
+
+            let discount_price = whitelist.discount_price.unwrap() as i64;
+            (
+                Some(mode),
+                Some(whitelist.presale),
+                Some(whitelist.mint.to_bytes().to_vec()),
+                Some(discount_price),
+            )
+        } else {
+            (None, None, None, None)
+        };
+
+    let (name, uri, hash) = if let Some(hidden_settings) = data.hidden_settings {
+        (
+            Some(hidden_settings.name),
+            Some(hidden_settings.uri),
+            Some(hidden_settings.hash.to_vec()),
+        )
+    } else {
+        (None, None, None)
+    };
+
+    let (expire_on_use, gatekeeper_network) = if let Some(gatekeeper) = data.gatekeeper {
+        (
+            Some(gatekeeper.expire_on_use),
+            Some(gatekeeper.gatekeeper_network.to_bytes().to_vec()),
+        )
+    } else {
+        (None, None)
+    };
```

Comment:
I guess this makes sense since all the database stuff wants these separate options.  I recall seeing the opposite version of `get_gatekeeper()` that takes separate options and puts them into a `Gatekeeper` struct.

Should both direction versions of all of the helpers be in the same helpers file?

Reply:
one is in api and one is in ingester. they need to be kept separate imo since they live in two different packages